File: scripts/themekit_release/release.rb
# frozen_string_literal: true
require 'digest'

# Release represents a single platform release
class Release
  attr_reader :platform, :file, :version, :full_name, :data
  attr_accessor :location

  def initialize(version: nil, platform: nil, filename: nil)
    @platform = platform
    @version = version
    load_file(filename)
  end

  def hexdigest
    @hexdigest ||= Digest::MD5.hexdigest(data)
  end

  def as_hash
    {
      name: platform,
      url: location,
      digest: hexdigest
    }
  end

  private

  def load_file(filename)
    file = File.open(filename, 'rb')
    @full_name = [version, platform, File.basename(file.path)].join('/')
    @data = file.read
    file.close
  end
end

File: cmd/tkrelease/release.go
package main

import (
	"crypto/md5"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/Shopify/themekit/kit"
)

var (
	builds = map[string]string{
		"darwin-amd64":  "theme",
		"darwin-386":    "theme",
		"linux-386":     "theme",
		"linux-amd64":   "theme",
		"windows-386":   "theme.exe",
		"windows-amd64": "theme.exe",
	}
)

func buildRelease(ver string, uploader *s3Uploader) (kit.Release, error) {
	var wg sync.WaitGroup
	wg.Add(len(builds))
	newRelease := kit.Release{Version: ver, Platforms: []kit.Platform{}}
	finished := make(chan bool, 1)
	errChan := make(chan error)
	platformChan := make(chan kit.Platform)

	for platformName, binName := range builds {
		go buildPlatform(ver, platformName, binName, uploader, platformChan, errChan, &wg)
	}

	go func() {
		wg.Wait()
		close(finished)
	}()

	for {
		select {
		case <-finished:
			return newRelease, nil
		case platform := <-platformChan:
			newRelease.Platforms = append(newRelease.Platforms, platform)
		case err := <-errChan:
			return kit.Release{}, err
		}
	}
}

func buildPlatform(ver, platformName, binName string, uploader *s3Uploader, platformChan chan kit.Platform, errChan chan error, wg *sync.WaitGroup) {
	defer wg.Done()

	f, err := os.Open(filepath.Join(distDir, platformName, binName))
	if err != nil {
		errChan <- err
		return
	}
	defer f.Close()

	data, err := ioutil.ReadAll(f)
	if err != nil {
		errChan <- err
		return
	}

	fullName := strings.Join([]string{ver, platformName, binName}, "/")
	url, err := uploader.file(fullName, f)
	if err != nil {
		errChan <- err
		return
	}

	platformChan <- kit.Platform{
		Name:   platformName,
		URL:    url,
		Digest: fmt.Sprintf("%x", md5.Sum(data)),
	}
}
